### DISCUSSION TITLE
Is it possible to automatically select a meow state when entering a new buffer which has a particular minor mode enabled

### DISCUSSION THREAD

**Discussion by smypf** [meow-edit/meow#674] [Q&A] [answered]
Meow provides the capability to allow a user to set the default state when a new buffer with a major mode is opened using the `meow-mode-state-list`. This is really useful as it allows for meow to get out of the way when in specific scenarios.

Is there functionality which allows for the same capabilities when a minor mode is present? Increasingly I am finding that there are minor modes where I experience friction as I want them to be in `insert` mode right away, rather than being in `normal` mode and having to enter `insert` mode manually.

Aside:
Thank you DogLooksGood for a fantastic package. Meow is fundamental to my editing experience. I'm quickly getting to the point where I struggle to operate other editors as my muscle memory is so well trained.

**Comment by Linerre** [comment]
As far as I know, Meow acts _almost_ the same in `insert-mode` as vanilla Emacs, except that you can escape to `normal-mode`.

But in your case, this would become awkward, because escaping to `normal-mode` nullifies your wish to stay in the `insert-mode` since the very first visit to a buffer with some (minor) mode.

How about this: add a hook to whatever minor mode you want to avoid `normal-mode`:
```
(add-hook 'some-minor-mode (lambda () (meow-mode -1)))
```

I once used this trick in some REPL buffers. But please keep in mind that I hooked this lambda to _major_ modes instead of minor ones.  Minor modes in Emacs are used like helpers for major modes as I understand it. So you may find it even more awkward to tweak Meow with minor modes.

**Reply by smypf** [reply]
Doesn't this disable meow completely? Editing is still part of the process and for me Meow is my optimal solution. 

I do admit that where available hooks would be a way of resolving the problem I'm facing.

**Reply by Linerre** [reply]
> Doesn't this disable meow completely? 

As I understand it, this only disables Meow in that specific buffer.

An even simpler approach would be
```
(add-hook 'some-minor-mode (lambda () (meow-insert-mode 1))
```

I didn't recommend this in the first place because according to your description, it seems you'd like to work in some buffers without Meow's getting in your way. But in the insertion mode, you can escape to normal mode if you (by accident) pressed Escape key, which doesn't seem to align with your expectation.

**Comment by DogLooksGood** [comment]
By meow's design, if you are in a buffer where text editing is allowed, then it should be initialized with NORMAL state, even though it's not a file buffer, i.e. REPL, Chat, etc.  One buffer can have multiple minor modes, but Meow has only one state at the same time. Could you tell about your use case? If it's some minor modes don't play well with meow, then it should be fixed on Meow side.

**Reply by smypf** [reply]
In my use case there are minor modes such as [`git-commit`](https://github.com/magit/magit/blob/main/lisp/git-commit.el#L630) and org-capture are uniquely used for specific scenarios. I would like to immediately start typing as I am already primed with an idea that needs to be written. I don't want to have Meow completely disabled since editing is still part of the writing process.

As @Linerre discusses above it would be possible to use the minor mode's hooks. It is a viable solution, but the responsibility is inverted. I think that Meow being responsible for which mode it selects when a new buffer is opened would mean centralise configuration. This would be difficult though as there can be multiple minor modes and this would introduce conflicts in which state to use by default.

If this functionality currently isn't available in Meow I see two paths forward:

1. Use hooks as suggested by @Linerre. This is entirely workable.
2. Implement this myself, eventually when I have time to get around to it :)

**Reply by Linerre** [reply, marked as answer]
> Implement this myself, eventually when I have time to get around to it 

There are two places where I think you may find helpful information:
1. When the major mode is `*Custom-mode*`, (entered with `M-x customize`), if you TAB to the search field, you can directly type in any characters, even now Meow is in the Motion mode.
2. As you may have noticed, whenever you are in the Minibuffer, you can type directly and Meow is disabled completely. 

But my knowledge of these two approaches are too shallow to provide any further hints on how you might be able to achieve exactly what you want.

**Reply by DogLooksGood** [reply]
I think solution from @Linerre is the right way to go. It's actually not a minor-mode problem, basically you can do anything with any hooks in Emacs, we should not duplicate nor specialize this feature in Meow.

**Reply by smypf** [reply]
Thank you @Linerre and @DogLooksGood for your assistance.
